### PR TITLE
Fix reordering semi joins

### PR DESCRIPTION
**Pull request by Tmonster**
The fuzzer has been producing a number of issue and they are related to our incorrect implementation of reordering semi joins. 

The problem is the following scenario. 
Suppose multiple LHS relations have a condition in a semi join. For example, suppose you have a join tree like ((A ⨝ B) ⋉ C). (sort of an example in test_4950.test)

And lets suppose the join condition looks like A.x = C.y AND B.x = C.z. Originally, each comparison is extracted as a potential filter. Doing so can result in reorderings that look like ( B ⨝ (A ⋉ C)). This reordering is invalid since all columns from C will be lost after the join with A, and the condition B.x = C.z will no longer be possible.

if we make a conjunction expressions and populate the left set and right set of the hyperedge with all the relations from the conditions in the conjunction expression, we can prevent these types of invalid reorderings. 

TL;DR, semi&anti join filters can only be applied when all relations on the left side are present.

Another issue we had was that we extracted all relations on the right side of the join as reorderable. This is also wrong because it can result in join trees where a relation A previously on the right side of the semi join is now above the semi join. This will result in more columns being projected as the columns from A are now in the result, when previously they would be lost after the semi join.

Another issue I encountered was when the left child of a semi join has a cross product. During relation extraction, the cross product filter isn't extracted. This is fine, because we would create cross products ourselves if we don't find a valid join tree. The problem, however, is that cross products were made between ALL relations, since we only reordered inner joins. With semi&anti joins, this should not be allowed, since the original positions of the relations relative to the semi/anti join needs to be respected.

### CODE REVIEW

**Comment by Tmonster** [comment]
There is a regression for tpch q2, I will investigate,

Also, I plan on cleaning up some of my comments soon

File: src/optimizer/join_order/cardinality_estimator.cpp [hunk 1]
```diff
@@ -107,6 +110,11 @@ void CardinalityEstimator::InitEquivalentRelations(const vector<unique_ptr<Filte
 			continue;
 		} else if (EmptyFilter(*filter)) {
 			continue;
+		} else if ((filter->join_type == JoinType::SEMI || filter->join_type == JoinType::ANTI) && !filter->left_set) {
+			// the right side is from some dummy scan and the comparison is
```
Comment (Tmonster) [inline src/optimizer/join_order/cardinality_estimator.cpp:114]: this condition can be removed as left_set and right_set are populated when relations are extracted from a semi join

**Comment by lnkuiper** [comment]
> TL;DR, semi&anti join filters can only be applied when all relations on the left side are present.

If the join conditions look like this instead: A.x = C.y AND B.x = C.y, we should still be able to reorder this as (B ⨝ (A ⋉ C)), since we can derive B.x = A.x. Does that still work?

> Another issue we had was that we extracted all relations on the right side of the join as reorderable. This is also wrong because it can result in join trees where a relation A previously on the right side of the semi join is now above the semi join. This will result in more columns being projected as the columns from A are now in the result, when previously they would be lost after the semi join.

I understand that, semantically, we should not be able to have access to these columns, but they shouldn't be referenced anyway. Also, won't these columns be removed by `RemoveUnusedColumns` and `ColumnLifetimeAnalyzer`?

> Another issue I encountered was when the left child of a semi join has a cross product. During relation extraction, the cross product filter isn't extracted. This is fine, because we would create cross products ourselves if we don't find a valid join tree. The problem, however, is that cross products were made between ALL relations, since we only reordered inner joins. With semi&anti joins, this should not be allowed, since the original positions of the relations relative to the semi/anti join needs to be respected.

I don't understand what you mean with the last sentence, as it sounds like semi/anti joins can't be reordered. Let's discuss later today.

**Comment by Tmonster** [comment]
> If the join conditions look like this instead: A.x = C.y AND B.x = C.y, we should still be able to reorder this as (B ⨝ (A ⋉ C)), since we can derive B.x = A.x. Does that still work?

Theoretically, yes this could work, but implementing would be out of scope for this PR. I would have to verify if it derived condition is another semi join or if it could be an inner join. Also, this would only work for semi joins, but not for anti joins.

> I understand that, semantically, we should not be able to have access to these columns, but they shouldn't be referenced anyway. Also, won't these columns be removed by RemoveUnusedColumns and ColumnLifetimeAnalyzer?

I guess I should have been more clear, we don't want to reorder relations in the RHS of the semi join to outside the semi join because of a join condition that might exist with that relation. If the RHS of a semi join has an inner join between tables A&B, we cannot reorder either A or B out of the semi join, because the columns of the other table are no longer available after the semi join. This would make the join condition that combines A&B invalid. In addition, reordering relation A out of the semi join could produce incorrect results for the semi join, as the RHS child potentially has more/less values in the column that is being semi joined on.

> I don't understand what you mean with the last sentence, as it sounds like semi/anti joins can't be reordered. Let's discuss later today.

This also wasn't super clear, but the children of a semi/anti join cannot move to the other side of the join. LHS children are allowed to be reordered to above the semi join, and v.v.

**Comment by lnkuiper** [comment]
> Theoretically, yes this could work, but implementing would be out of scope for this PR. I would have to verify if it derived condition is another semi join or if it could be an inner join. Also, this would only work for semi joins, but not for anti joins.

Good idea to leave it out-of-scope for this PR. Also since it only works for SEMI and not ANTI might not be worth the effort.

Thanks for the explanations!

File: src/include/duckdb/optimizer/join_order/query_graph_manager.hpp [hunk 2]
```diff
@@ -48,13 +48,16 @@ class FilterInfo {
 
 public:
 	unique_ptr<Expression> filter;
-	JoinRelationSet &set;
+	optional_ptr<JoinRelationSet> set;
```
Comment (lnkuiper) [inline src/include/duckdb/optimizer/join_order/query_graph_manager.hpp:51]: Nitpick: could this be a `reference<JoinRelationSet>` instead? This reference can be changed to refer to something else later on, which I think you're doing here. The only difference with `optional_ptr` is that it must always be initialized and, therefore, cannot be `nullptr`.

File: test/optimizer/joins/delim_join_dont_explode.test_slow [hunk 3]
```diff
@@ -62,7 +63,7 @@ exists(
 )
 order by bt.id
 ----
-physical_plan	<REGEX>:.*HASH_JOIN.*SEQ_SCAN.*DELIM_SCAN.*SEQ_SCAN.*
+physical_plan	<REGEX>:.*HASH_JOIN.*DELIM_SCAN.*SEQ_SCAN.*SEQ_SCAN.*
```
Comment (lnkuiper) [inline test/optimizer/joins/delim_join_dont_explode.test_slow:66]: Why has this plan changed?

This the the plan on `main`:
```
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Projections:       │
│__internal_decompress_integ│
│     ral_integer(#0, 0)    │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│   Estimated Cardinality:  │
│            1000           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│          ORDER_BY         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         Order By:         │
│         bt.id ASC         │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Projections:       │
│__internal_compress_integra│
│     l_usmallint(#0, 0)    │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│   Estimated Cardinality:  │
│            1000           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│      RIGHT_DELIM_JOIN     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         Join Type:        │
│         RIGHT_SEMI        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Conditions:        │
│ id IS NOT DISTINCT FROM id├──────────────┐
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
│   Estimated Cardinality:  │              │
│            1000           │              │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
│        Delim Index:       │              │
│             1             │              │
└─────────────┬─────────────┘              │
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│         SEQ_SCAN          ││         HASH_JOIN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Stringified:       ││         Join Type:        │
│         big_table         ││         RIGHT_SEMI        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Projections:       ││        Conditions:        ├────────────────────────────────────────────────────────────────────────┐
│             id            ││ id IS NOT DISTINCT FROM id│                                                                        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                        │
│   Estimated Cardinality:  ││   Estimated Cardinality:  │                                                                        │
│            1000           ││            1000           │                                                                        │
└───────────────────────────┘└─────────────┬─────────────┘                                                                        │
                             ┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
                             │         PROJECTION        │                                                          │         DUMMY_SCAN        │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          │                           │
                             │        Projections:       │                                                          │                           │
                             │             id            │                                                          │                           │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          │                           │
                             │   Estimated Cardinality:  │                                                          │                           │
                             │             50            │                                                          │                           │
                             └─────────────┬─────────────┘                                                          └───────────────────────────┘
                             ┌─────────────┴─────────────┐
                             │         HASH_JOIN         │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │         Join Type:        │
                             │           INNER           │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │        Conditions:        │
                             │    id = fk_to_medium_2    │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │         Build Min:        ├──────────────┐
                             │             99            │              │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
                             │         Build Max:        │              │
                             │             99            │              │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
                             │   Estimated Cardinality:  │              │
                             │             50            │              │
                             └─────────────┬─────────────┘              │
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │         SEQ_SCAN          ││         HASH_JOIN         │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │        Stringified:       ││         Join Type:        │
                             │          medium_2         ││           INNER           │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │        Projections:       ││        Conditions:        │
                             │             id            ││ (id % 7) = (fk_to_big % 7)├──────────────┐
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
                             │          Filters:         ││   Estimated Cardinality:  │              │
                             │ id>=99 AND id IS NOT NULL ││            500            │              │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││                           │              │
                             │   Estimated Cardinality:  ││                           │              │
                             │            100            ││                           │              │
                             └───────────────────────────┘└─────────────┬─────────────┘              │
                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                          │         DELIM_SCAN        ││         SEQ_SCAN          │
                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                                                          │        Delim Index:       ││        Stringified:       │
                                                          │             1             ││          medium_1         │
                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                                                          │   Estimated Cardinality:  ││        Projections:       │
                                                          │            500            ││       fk_to_medium_2      │
                                                          │                           ││         fk_to_big         │
                                                          │                           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                                                          │                           ││          Filters:         │
                                                          │                           ││   fk_to_medium_2<=99 AND  │
                                                          │                           ││ fk_to_medium_2 IS NOT NULL│
                                                          │                           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                                                          │                           ││   Estimated Cardinality:  │
                                                          │                           ││            100            │
                                                          └───────────────────────────┘└───────────────────────────┘
```
The REGEX tests specifically this part of the query:
```
┌─────────────┴─────────────┐
│         HASH_JOIN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         Join Type:        │
│           INNER           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Conditions:        │
│    id = fk_to_medium_2    │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         Build Min:        ├──────────────┐
│             99            │              │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
│         Build Max:        │              │
│             99            │              │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
│   Estimated Cardinality:  │              │
│             50            │              │
└─────────────┬─────────────┘              │
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│         SEQ_SCAN          ││         HASH_JOIN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Stringified:       ││         Join Type:        │
│          medium_2         ││           INNER           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        Projections:       ││        Conditions:        │
│             id            ││ (id % 7) = (fk_to_big % 7)├──────────────┐
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │
│          Filters:         ││   Estimated Cardinality:  │              │
│ id>=99 AND id IS NOT NULL ││            500            │              │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││                           │              │
│   Estimated Cardinality:  ││                           │              │
│            100            ││                           │              │
└───────────────────────────┘└─────────────┬─────────────┘              │
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │         DELIM_SCAN        ││         SEQ_SCAN          │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │        Delim Index:       ││        Stringified:       │
                             │             1             ││          medium_1         │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │   Estimated Cardinality:  ││        Projections:       │
                             │            500            ││       fk_to_medium_2      │
                             │                           ││         fk_to_big         │
                             │                           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │                           ││          Filters:         │
                             │                           ││   fk_to_medium_2<=99 AND  │
                             │                           ││ fk_to_medium_2 IS NOT NULL│
                             │                           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │                           ││   Estimated Cardinality:  │
                             │                           ││            100            │
                             └───────────────────────────┘└───────────────────────────┘
```
Which is just 2 inner joins. Shouldn't this part of the plan stay the same?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Thanks for the PR. I left some comments:

File: test/optimizer/joins/delim_join_dont_explode.test_slow (hunk 3 above)
Comment (lnkuiper) [reply test/optimizer/joins/delim_join_dont_explode.test_slow:66]: We discussed this and want to merge the PR as is since the join order does not affect correctness, while the semi/anti join ordering fix does affect correctness. This should be investigated in a future PR.

**Review by lnkuiper — APPROVED** [review]
Thanks for the changes, this looks good to me now. As mentioned the delim scan ordering should be investigated in a future PR.

**Comment by Mytherin** [comment]
Thanks!
